pkg/statsutil/stats.go

```diff
@@ -0,0 +1,380 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package statsutil
+
+import (
+	"fmt"
+	"runtime"
+	"sync"
+	"time"
+
+	wstats "github.com/Microsoft/hcsshim/cmd/containerd-shim-runhcs-v1/stats"
+	v1 "github.com/containerd/cgroups/stats/v1"
+	v2 "github.com/containerd/cgroups/v2/stats"
+	units "github.com/docker/go-units"
+)
+
+// StatsEntry represents the statistics data collected from a container
+type StatsEntry struct {
+	Container        string
+	Name             string
+	ID               string
+	CPUPercentage    float64
+	Memory           float64
+	MemoryLimit      float64
+	MemoryPercentage float64
+	NetworkRx        float64
+	NetworkTx        float64
+	BlockRead        float64
+	BlockWrite       float64
+	PidsCurrent      uint64
+	IsInvalid        bool
+}
+
+// FormattedStatsEntry represents a formatted StatsEntry
+type FormattedStatsEntry struct {
+	Name     string
+	ID       string
+	CPUPerc  string
+	MemUsage string
+	MemPerc  string
+	NetIO    string
+	BlockIO  string
+	PIDs     string
+}
+
+// Stats represents an entity to store containers statistics synchronously
+type Stats struct {
+	mutex sync.Mutex
```

Comment:
Would an RWMutex be preferred here? Seems it would benefit the Get methods

Reply:
thanks, fixed